src/securemul.rs

```diff
@@ -0,0 +1,430 @@
+#[derive(Debug)]
+pub struct SecureMul<F> {
+    index: u128,
+    a_share: ReplicatedSecretSharing<F>,
+    b_share: ReplicatedSecretSharing<F>,
+}
+
+/// A message sent by each helper when they've multiplied their own shares
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct DValue {
+    index: u128,
+    d: u128,
+}
+
+/// Context used by each helper to perform computation. Currently they need access to shared
+/// randomness generator (PRSS) and communication trait to send messages to each other.
+/// Eventually when we have more than one protocol, this should be lifted to its own module
+#[derive(Debug)]
+pub struct ProtocolContext<'a, R> {
+    pub participant: &'a Participant,
+    pub helper_ring: &'a R,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error(
+        "Shares calculated by peer used different index {their_index} than expected {my_index}"
+    )]
+    IndexMismatch { my_index: u128, their_index: u128 },
+}
+
+impl<F: Field> SecureMul<F> {
+    /// Executes the secure multiplication on the MPC helper side. Each helper will proceed with
+    /// their part, eventually producing 2/3 shares of the product and that is what this function
+    /// returns.
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    pub async fn execute<R: Ring>(
+        self,
+        ctx: &ProtocolContext<'_, R>,
+    ) -> Result<ReplicatedSecretSharing<F>, BoxError> {
+        // generate shared randomness.
+        let (s0, s1) = ctx.participant.generate_fields(self.index);
+
+        // compute the value (d_i) we want to send to the right helper (i+1)
+        let (a0, a1) = self.a_share.as_tuple();
+        let (b0, b1) = self.b_share.as_tuple();
+        let right_d: F = a0 * b1 + a1 * b0 - s0;
+
+        // this ugliness is needed just to convert Field to u128. There are better ways to do it
+        // and there is a PR open to make it easier
+        let right_d: <F as Field>::Integer = right_d.into();
+        let right_d: u128 = right_d.into();
+
+        // notify helper on the right that we've computed our value
+        ctx.helper_ring
+            .send(
+                HelperAddr::Right,
+                DValue {
+                    d: right_d,
+                    index: self.index,
+                },
+            )
+            .await?;
+
+        // Sleep until helper on the left sends us their (d_i-1) value
+        let DValue {
+            d: left_d,
+            index: left_index,
+        } = ctx.helper_ring.receive(HelperAddr::Left).await?;
+
+        // sanity check to make sure they've computed it using the same seed
+        if left_index == self.index {
+            // now we are ready to construct the result - 2/3 secret shares of a * b.
+            let lhs = a0 * b0 + F::from(left_d) + s0;
+            let rhs = a1 * b1 + s1 + F::from(right_d);
```

Comment:
Just a nit - but can you put the terms in the same order?

That is, make the second line:

  let rhs = a1 * b1 + F::from(right_d) + s1;